google/cloud/storage/examples/storage_object_samples.cc

```diff
@@ -322,6 +322,72 @@ void WriteLargeObject(google::cloud::storage::Client client, int& argc,
   (std::move(client), bucket_name, object_name, object_size_in_MiB);
 }
 
+void StartResumableUpload(google::cloud::storage::Client client, int& argc,
+                          char* argv[]) {
+  if (argc != 3) {
+    throw Usage{"start-resumable-upload <bucket-name> <object-name>"};
+  }
+  auto bucket_name = ConsumeArg(argc, argv);
+  auto object_name = ConsumeArg(argc, argv);
+
+  //! [start resumable upload]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string bucket_name, std::string object_name) {
+    gcs::ObjectWriteStream stream = client.WriteObject(
+        bucket_name, object_name, gcs::NewResumableUploadSession());
+    std::cout << "Created resumable upload: " << stream.resumable_session_id()
+              << std::endl;
+    // As it is customary in C++, the destructor automatically closes the
+    // stream, that would finish the upload and create the object. For this
+    // example we want to restore the session as-if the application had crashed,
+    // where no destructors get called.
+    stream << "This data will not get uploaded, it is too small" << std::endl;
+    std::move(stream).Suspend();
+  }
+  //! [start resumable upload]
+  (std::move(client), bucket_name, object_name);
+}
+
+void ResumeResumableUpload(google::cloud::storage::Client client, int &argc,
+                           char **argv) {
+  if (argc != 4) {
+    throw Usage{
+        "resume-resumable-upload <bucket-name> <object-name> <session-id>"};
+  }
+  auto bucket_name = ConsumeArg(argc, argv);
+  auto object_name = ConsumeArg(argc, argv);
+  auto session_id = ConsumeArg(argc, argv);
+
+  //! [resume resumable upload]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string bucket_name, std::string object_name,
+     std::string session_id) {
+    gcs::ObjectWriteStream stream =
+        client.WriteObject(bucket_name, object_name,
+                           gcs::RestoreResumableUploadSession(session_id));
```

Comment:
woah, this is awesome! A user was asking for this in the Go client library.

Reply:
Thanks, does this mean you are Okay with the changes?

Reply:
Ah.... apologies, I didn't click update for my comment.

Question, how does a user determine the offset when resuming the upload session? Does the library help query the service?

Reply:
Yes, the library automatically queries the service to find out the next expected byte.

Reply:
Thanks: Request: add a comment stating that the byte offset is queried when a resumable session id is provided.

Other than that LGTM. I appreciate your patience!